### DISCUSSION TITLE
Unable to run Synthea 2.7.0

### DISCUSSION THREAD

**Discussion by fabkury** [synthetichealth/synthea#1053] [Q&A] [answered]
I am trying to use https://github.com/OHDSI/ETL-Synthea, and it only supports version 2.7.0.

When I try to run synthea-with-dependencies.jar (from the 2.7.0 release), I get obscure errors like:

java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.concurrent.atomic.AtomicLong java.util.Random.seed accessible: module java.base does not "opens java.util" to unnamed module @ea1a8d5

From my debugging it seems this has to do with the Java version. I tried installing different versions (JDK 8, JDK 11, JDK 17...), then  using specifically those java.exe's to run synthea-with-dependencies.jar. I also tried building Synthea from sources, but when I follow the instructions, I get the error "Unsupported class file major version 61."

I just need to run Synthea 2.7.0. Any suggestions?

**Comment by fdefalco** [comment]
I believe I worked around this issue by using the following flags:

java -jar --illegal-access=permit

**Comment by jawalonoski** [comment]
Did fdefalco's suggested solution work?

You should be able to use Java 17. You can check the version you are actually running using `java -version`

According to the internet, if you are using your own Gradle environment rather than the local one in this repository, then your own environment could also be configured to use a different version of Gradle or Java. If that applies to you, that might be worth exploring.

**Comment by fabkury** [comment]
I tried, but couldn't get it to work. f I type
`java -jar --illegal-access=permit`
Java says:
`Java HotSpot(TM) 64-Bit Server VM warning: Ignoring option --illegal-access=permit; support was removed in 17.0`

If I move the `--illegal-access=permit` to the end of the command line, and add other restrictions, I get different, equally abstruse Java errors. I tried both Java 18 and Java 17, as well as Java 8. (!!)

This is the "least scary" error I was able to arrive at:
```
C:\Program Files\Java\jdk-17.0.3\bin> .\java.exe -jar "C:\[redacted]\synthea\synthea-with-dependencies270.jar" California -c "C:\[redacted]\synthea\synthea_settings.ini" -p 1000 --exporter.fhir.transaction_bundle false --exporter.csv.export true --illegal-access=permit
Scanned 66 modules and 67 submodules.
Loading submodule modules/breast_cancer/tnm_diagnosis.json
java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.concurrent.atomic.AtomicLong java.util.Random.seed accessible: module java.base does not "opens java.util" to unnamed module @ea1a8d5
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
        at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:178)
        at java.base/java.lang.reflect.Field.setAccessible(Field.java:172)
        at com.google.gson.internal.reflect.UnsafeReflectionAccessor.makeAccessible(UnsafeReflectionAccessor.java:44)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.getBoundFields(ReflectiveTypeAdapterFactory.java:159)
        at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory.create(ReflectiveTypeAdapterFactory.java:102)
        at com.google.gson.Gson.getAdapter(Gson.java:458) 
```
_(followed by barrage of Java errors)_

My plan now is to try to dig into how to get Synthea to work, without the precompiled JAR.

Thanks for the attention!!

**Comment by eedrummer** [comment, marked as answer]
For the record, Synthea support for Java 17 was added after 2.7.0. From a quick scan of closed PRs, I don't see anything that would immediately address your issue, but if you can use a newer version, I'd recommend that you do.

When Synthea reads in modules, it uses a library called Gson to read in the module files and set that information on Java objects in the running environment. The permissions on what could be set were changed from Java 16 -> Java 17 and we did see it break stuff. I'm surprised to hear that it broke with Java 8. It may be that the settings you are using a triggering a path in the code we haven't hit yet.

**Comment by fabkury** [comment]
Thanks, @eedrummer!

**I was finally able to run Synthea 2.7.0!** Thanks!

What I did was:
1. Install Java 11 JDK -- I don't think there's anything special about 11, _maybe_ the matter is to be pre-17 (just a guess!).
2. Go to Windows system environment variables, make sure the first Java path on the list is the Java 11 JDK bin folder.
3. Open Windows Terminal, use `where java` to double-check the first path on the list is Java 11 JDK. (It should be)
4. Run Synthea 2.7.0 regularly.

My final guess is that the problem was that I was installing older Javas (to run Synthea 2.7.0), but then merely using the old java.exe to run 
Synthea, instead of having the Windows system properly configured to always use the older Java. But again, this is just a guess.

Kind regards, thanks for the great work!
